### PR TITLE
scheduler: update the scheduler interface and cache methods to use contextual logging

### PR DESCRIPTION
**Pull request by mengjiao-liu**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Update the scheduler interface and cache methods to use contextual logging.

#### Which issue(s) this PR fixes:

Part of https://github.com/kubernetes/kubernetes/issues/91633#issuecomment-1467645065

#### Special notes for your reviewer:
Update the interface, cache, and some of the files that call these methods.
Because these methods are called from many places and cannot be subdivided, this PR may be large.

#### Does this PR introduce a user-facing change?

```release-note
Update the scheduler interface and cache methods to use [contextual logging](https://k8s.io/docs/concepts/cluster-administration/system-logs/#contextual-logging).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3077-contextual-logging
```

### CODE REVIEW

File: pkg/scheduler/internal/cache/interface.go [hunk 1]
```diff
@@ -95,21 +96,21 @@ type Cache interface {
 
 	// AddNode adds overall information about node.
 	// It returns a clone of added NodeInfo object.
-	AddNode(node *v1.Node) *framework.NodeInfo
+	AddNode(logger klog.Logger, node *v1.Node) *framework.NodeInfo
 
 	// UpdateNode updates overall information about node.
 	// It returns a clone of updated NodeInfo object.
-	UpdateNode(oldNode, newNode *v1.Node) *framework.NodeInfo
+	UpdateNode(logger klog.Logger, oldNode, newNode *v1.Node) *framework.NodeInfo
```
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/internal/cache/interface.go:103]: For contextual logging there's no way around adding either klog.Logger or a context to the function.

The guidance is to use klog.Logger when the function doesn't need a context for something else.

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/internal/cache/interface.go:103]: hm... maybe we should avoid such logging in internal. But handle it outside.

[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/internal/cache/interface.go:103]: Also cc @alculquicondor @Huang-Wei who reviewed the contextual logging before.

[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/internal/cache/interface.go:103]: A logger per instance is indeed problematic. When calling methods of that instance, the context in which the method gets called would be ignored when using that logger.

**Comment by alculquicondor** [comment]
@kerthcet lgtm?

**Comment by mengjiao-liu** [comment]
Conflicts due to other PR merged need to be rebased again.

File: pkg/scheduler/internal/cache/interface.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/interface.go:103]: > A logger per instance is indeed problematic. When calling methods of that instance, the context in which the method gets called would be ignored when using that logger.

I don't quite understand this sentence, how should I modify it? @pohly

**Comment by alculquicondor** [comment]
will review after rebases.
Also probably better to wait for #116842 to merge first.

File: pkg/scheduler/framework/interface.go [hunk 2]
```diff
@@ -704,11 +705,11 @@ func (ni *NominatingInfo) Mode() NominatingMode {
 type PodNominator interface {
 	// AddNominatedPod adds the given pod to the nominator or
 	// updates it if it already exists.
-	AddNominatedPod(pod *PodInfo, nominatingInfo *NominatingInfo)
+	AddNominatedPod(logger klog.Logger, pod *PodInfo, nominatingInfo *NominatingInfo)
```
Comment (kerthcet) [inline pkg/scheduler/framework/interface.go:708]: I think pass in context is more expandable for interface.

File: pkg/scheduler/internal/cache/cache.go [hunk 3]
```diff
@@ -545,17 +548,18 @@ func (cache *cacheImpl) UpdatePod(oldPod, newPod *v1.Pod) error {
 	// before Update event, in which case the state would change from Assumed to Added.
 	if cache.assumedPods.Has(key) {
 		return fmt.Errorf("assumed pod %v(%v) should not be updated", key, klog.KObj(oldPod))
+
```
Comment (kerthcet) [inline pkg/scheduler/internal/cache/cache.go:551]: revert

File: pkg/scheduler/internal/cache/cache.go [hunk 4]
```diff
@@ -569,15 +573,15 @@ func (cache *cacheImpl) RemovePod(pod *v1.Pod) error {
...
 	if currState.pod.Spec.NodeName != pod.Spec.NodeName {
-		klog.ErrorS(nil, "Pod was added to a different node than it was assumed", "podKey", key, "pod", klog.KObj(pod), "assumedNode", klog.KRef("", pod.Spec.NodeName), "currentNode", klog.KRef("", currState.pod.Spec.NodeName))
+		logger.Error(nil, "Pod was added to a different node than it was assumed", "podKey", key, "pod", klog.KObj(pod), "assumedNode", klog.KRef("", pod.Spec.NodeName), "currentNode", klog.KRef("", currState.pod.Spec.NodeName))
 		if pod.Spec.NodeName != "" {
 			// An empty NodeName is possible when the scheduler misses a Delete
 			// event and it gets the last known state from the informer cache.
-			klog.ErrorS(nil, "scheduler cache is corrupted and can badly affect scheduling decisions")
-			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
+			logger.Error(nil, "scheduler cache is corrupted and can badly affect scheduling decisions")
+			os.Exit(1)
```
Comment (kerthcet) [inline pkg/scheduler/internal/cache/cache.go:581]: It seems incompatible, but we do not have similar functions like `flushAndExit` under contextual logger, right?

File: pkg/scheduler/internal/cache/cache.go [hunk 5]
```diff
@@ -751,17 +753,18 @@ func (cache *cacheImpl) cleanupAssumedPods(now time.Time) {
 	for key := range cache.assumedPods {
 		ps, ok := cache.podStates[key]
 		if !ok {
-			klog.ErrorS(nil, "Key found in assumed set but not in podStates, potentially a logical error")
+			logger.Error(nil, "Key found in assumed set but not in podStates, potentially a logical error")
 			klog.FlushAndExit(klog.ExitFlushTimeout, 1)
+			// os.Exit(1)
```
Comment (kerthcet) [inline pkg/scheduler/internal/cache/cache.go:758]: ?

File: pkg/scheduler/internal/cache/interface.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/internal/cache/interface.go:103]: Then I think no need to use contextual logging here.

File: pkg/scheduler/internal/cache/cache.go (hunk 4 above)
Comment (kerthcet) [reply pkg/scheduler/internal/cache/cache.go:581]: If we use klog here, then no need to update.

File: pkg/scheduler/eventhandlers.go [hunk 6]
```diff
@@ -43,9 +43,10 @@ import (
 )
 
 func (sched *Scheduler) onStorageClassAdd(obj interface{}) {
+	logger := sched.logger
 	sc, ok := obj.(*storagev1.StorageClass)
 	if !ok {
-		klog.ErrorS(nil, "Cannot convert to *storagev1.StorageClass", "obj", obj)
+		logger.Error(nil, "Cannot convert to *storagev1.StorageClass", "obj", obj)
```
Comment (kerthcet) [inline pkg/scheduler/eventhandlers.go:49]: Also think klog is enough. Then maybe we can remove logger from scheduler.

**Review by kerthcet — COMMENTED** [review]
Several comments, I haven't review the whole codebase yet, one point I want to highlight is if not necessary to use contextual logging, we can use klog directly, like cache and eventhandler. Agree? cc @pohly @alculquicondor

File: pkg/scheduler/internal/cache/cache.go (hunk 4 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/cache.go:581]: @pohly  https://github.com/kubernetes/klog/pull/306 It looks like contextual logging now supports flush(klog.FlushAndExit), right?

File: pkg/scheduler/internal/cache/cache.go (hunk 5 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/cache.go:758]: The previous PR legacy, sorry, I did not notice this, next time I will modify together.

File: pkg/scheduler/internal/cache/cache.go (hunk 3 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/cache.go:551]: Okay. Next time modify it together.

File: pkg/scheduler/framework/interface.go (hunk 2 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/interface.go:708]: Pass the context is also ok.

File: pkg/scheduler/internal/cache/cache.go (hunk 4 above)
Comment (pohly) [reply pkg/scheduler/internal/cache/cache.go:581]: Yes. Use `logger.Error` and `klog.FlushAndExit` here.

File: pkg/scheduler/eventhandlers.go (hunk 6 above)
Comment (kerthcet) [reply pkg/scheduler/eventhandlers.go:49]: Can you explain more about this? Why unit/integration tests will impact this, the logger was inherited from scheduler, it has no context of the tests. 

And since we added no names or values here, so the log output should be same.

**Comment by mengjiao-liu** [comment]
Tests failed, continue to solve tomorrow.

File: pkg/scheduler/framework/interface.go (hunk 2 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/interface.go]: > I think pass in context is more expandable for interface.

I agree that passing context in an interface is more extensible, but passing context has one problem, as explained below.

I tried passing the context in the following interface:
```
type Cache interface
type SchedulingQueue interface 
type PodNominator interface
```

I found that eventually I needed to modify the resource events handler, such as


			Handler: cache.ResourceEventHandlerFuncs{
				AddFunc:    sched.addPodToSchedulingQueue,
				UpdateFunc: sched.updatePodInSchedulingQueue,
				DeleteFunc: sched.deletePodFromSchedulingQueue,
				AddFunc: func(obj interface{}) {
					sched.addPodToSchedulingQueue(ctx, obj)
				},
				UpdateFunc: func(oldObj, newObj interface{}) {
					sched.updatePodInSchedulingQueue(ctx, oldObj, newObj)
				},
				DeleteFunc: func(obj interface{}) {
					sched.deletePodFromSchedulingQueue(ctx, obj)
				},
			},



The context is always going to be the same in the same scheduler instance. Better look up the logger once and pass that.

In other words,  each time a new event is received, we need to get the logger from the context, which is more expensive than getting the `sched.logger` directly at one time.

Compare details about passing the logger to passing the context: https://github.com/kubernetes/kubernetes/compare/234db6b0479815ce97666eedf26784c874461c25..4b885627199e83cc0a326f0a99189c1759d5bb3f

@alculquicondor  @kerthcet 

ps: I can also change back to pass logger if needed.

File: pkg/scheduler/internal/cache/interface.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/interface.go]: > A logger per instance is indeed problematic. When calling methods of that instance, the context in which the method gets called would be ignored when using that logger.

Does this mean that if we pass the context instead of `schd.logger` we can solve the problem? @pohly

**Comment by mengjiao-liu** [comment]
Pass context parameter in the interface instead of logger parameter.

File: pkg/scheduler/internal/cache/interface.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pohly) [reply pkg/scheduler/internal/cache/interface.go]: Passing either a `Logger` or `Context` into a function is both fine. Which one is better depends on whether the function will ever need a `Context` - if yes, use that, if no, use `Logger`.

A field in a struct is problematic. It leads to confusion when other methods of that struct need to do logging and take a `Context`. Therefore I prefer to avoid them and instead add `Logger` parameters.

For this particular example, `UpdateNode(logger klog.Logger, oldNode, newNode *v1.Node) *framework.NodeInfo` seems fine to me.

File: pkg/scheduler/framework/interface.go (hunk 2 above)
Comment (mengjiao-liu) [reply pkg/scheduler/framework/interface.go]: > I think pass in context is more expandable for interface.

I agree that passing context in an interface is more extensible, but passing context has one problem, as explained below.

I tried passing the context in the following interface:
```
type Cache interface
type SchedulingQueue interface 
type PodNominator interface
```

I found that eventually I needed to modify the resource event handlers, such as


			Handler: cache.ResourceEventHandlerFuncs{
				AddFunc:    sched.addPodToSchedulingQueue,
				UpdateFunc: sched.updatePodInSchedulingQueue,
				DeleteFunc: sched.deletePodFromSchedulingQueue,
				AddFunc: func(obj interface{}) {
					sched.addPodToSchedulingQueue(ctx, obj)
				},
				UpdateFunc: func(oldObj, newObj interface{}) {
					sched.updatePodInSchedulingQueue(ctx, oldObj, newObj)
				},
				DeleteFunc: func(obj interface{}) {
					sched.deletePodFromSchedulingQueue(ctx, obj)
				},
			},



The context is always going to be the same in the same scheduler instance. Better look up the logger once and pass that.

In other words,  each time a new event is received, we need to get the logger from the context, which is more expensive than passing logger directly at one time.

Compare details about passing the logger to passing the context: https://github.com/kubernetes/kubernetes/compare/234db6b0479815ce97666eedf26784c874461c25..4b885627199e83cc0a326f0a99189c1759d5bb3f

ref to：https://github.com/kubernetes/kubernetes/pull/116849#discussion_r1198655475

@alculquicondor  @kerthcet 

ps: I can also change back to pass logger if needed.

Comment (alculquicondor) [reply pkg/scheduler/framework/interface.go]: Yes, pass logger wherever possible, as it's more performant.

**Comment by alculquicondor** [comment]
/hold
for #118012

File: pkg/scheduler/framework/interface.go (hunk 2 above)
Comment (kerthcet) [reply pkg/scheduler/framework/interface.go]: As @alculquicondor , let's revert to use logger, although it's a little ugly. And sorry for that.

Comment (mengjiao-liu) [reply pkg/scheduler/framework/interface.go:708]: Don't be sorry, just for the better~

I have reverted to using logger.

**Comment by mengjiao-liu** [comment]
/hold cancel 
https://github.com/kubernetes/kubernetes/pull/118012 has been merged.

File: pkg/scheduler/internal/cache/cache_test.go [hunk 7]
```diff
@@ -321,26 +327,30 @@ func TestExpirePod(t *testing.T) {
 		},
 	}
 
-	for _, tc := range tests {
-		t.Run(tc.name, func(t *testing.T) {
-			cache := newCache(tc.ttl, time.Second, nil)
+	for i, tt := range tests {
+		t.Run(fmt.Sprintf("case_%d", i), func(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/cache/cache_test.go:331]:
```suggestion
		t.Run(tt.name, func(t *testing.T) {
```

Comment (alculquicondor) [reply pkg/scheduler/internal/cache/cache_test.go:331]: Although I think we are using more `tc` these days.

File: pkg/scheduler/internal/cache/cache_test.go [hunk 8]
```diff
@@ -383,22 +393,25 @@ func TestAddPodWillConfirm(t *testing.T) {
 
 	for i, tt := range tests {
 		t.Run(fmt.Sprintf("case_%d", i), func(t *testing.T) {
```
Comment (alculquicondor) [inline pkg/scheduler/internal/cache/cache_test.go:395]: In a follow up, we can remove the test table, as there is only one test. Or add more tests and a name.

And similar for other cases below.

File: pkg/scheduler/internal/cache/node_tree.go [hunk 9]
```diff
@@ -36,24 +36,24 @@ type nodeTree struct {
...
 // addNode adds a node and its corresponding zone to the tree. If the zone already exists, the node
 // is added to the array of nodes in that zone.
-func (nt *nodeTree) addNode(n *v1.Node) {
+func (nt *nodeTree) addNode(logger klog.Logger, n *v1.Node) {
 	zone := utilnode.GetZoneKey(n)
 	if na, ok := nt.tree[zone]; ok {
 		for _, nodeName := range na {
 			if nodeName == n.Name {
-				klog.InfoS("Node already exists in the NodeTree", "node", klog.KObj(n))
+				logger.Info("Node already exists in the NodeTree", "node", klog.KObj(n))
```
Comment (alculquicondor) [inline pkg/scheduler/internal/cache/node_tree.go:56]:
```suggestion
				logger.Info("Did not add to the NodeTree because it already exists", "node", klog.KObj(n))
```

File: pkg/scheduler/internal/cache/node_tree.go [hunk 10]
```diff
@@ -76,13 +76,13 @@ func (nt *nodeTree) removeNode(n *v1.Node) error {
...
 				}
-				klog.V(2).InfoS("Removed node in listed group from NodeTree", "node", klog.KObj(n), "zone", zone)
+				logger.V(2).Info("Removed node in listed group from NodeTree", "node", klog.KObj(n), "zone", zone)
 				nt.numNodes--
 				return nil
 			}
 		}
 	}
-	klog.ErrorS(nil, "Node in listed group was not found", "node", klog.KObj(n), "zone", zone)
+	logger.Error(nil, "Node in listed group was not found", "node", klog.KObj(n), "zone", zone)
```
Comment (alculquicondor) [inline pkg/scheduler/internal/cache/node_tree.go:85]:
```suggestion
	logger.Error(nil, "Did not remove Node in NodeTree because it was not found", "node", klog.KObj(n), "zone", zone)
```

File: pkg/scheduler/schedule_one.go [hunk 11]
```diff
@@ -320,6 +324,7 @@ func (sched *Scheduler) skipPodSchedule(fwk framework.Framework, pod *v1.Pod) bo
 	// during its previous scheduling cycle but before getting assumed.
 	isAssumed, err := sched.Cache.IsAssumedPod(pod)
 	if err != nil {
+		// TODO: pass ctx into a revised HandleError
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:327]: Is there an issue for this? If so, please reference `TODO(#1234)`
 Also, does it need to be a context or can it be a logger?

File: pkg/scheduler/schedule_one.go [hunk 12]
```diff
@@ -152,9 +154,9 @@ func (sched *Scheduler) schedulingCycle(
 		msg := status.Message()
 		fitError.Diagnosis.PostFilterMsg = msg
 		if status.Code() == framework.Error {
-			klog.ErrorS(nil, "Status after running PostFilter plugins for pod", "pod", klog.KObj(pod), "status", msg)
+			logger.Error(nil, "Status after running PostFilter plugins for pod", "pod", klog.KObj(pod), "status", msg)
 		} else {
-			klog.V(5).InfoS("Status after running PostFilter plugins for pod", "pod", klog.KObj(pod), "status", msg)
+			logger.Error(nil, "Status after running PostFilter plugins for pod", "pod", klog.KObj(pod), "status", msg)
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:159]: why Error?

File: pkg/scheduler/schedule_one.go [hunk 13]
```diff
@@ -663,11 +670,11 @@ func prioritizeNodes(
...
 
 	// Additional details logged at level 10 if enabled.
-	klogV := klog.V(10)
-	if klogV.Enabled() {
+	loggerVTen := logger.V(10)
+	if loggerVTen.Enabled() {
 		for _, nodeScore := range nodesScores {
 			for _, pluginScore := range nodeScore.Scores {
-				klogV.InfoS("Plugin scored node for pod", "pod", klog.KObj(pod), "plugin", pluginScore.Name, "node", nodeScore.Name, "score", pluginScore.Score)
+				logger.V(10).Info("Plugin scored node for pod", "pod", klog.KObj(pod), "plugin", pluginScore.Name, "node", nodeScore.Name, "score", pluginScore.Score)
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:677]:
```suggestion
				loggerVTen.Info("Plugin scored node for pod", "pod", klog.KObj(pod), "plugin", pluginScore.Name, "node", nodeScore.Name, "score", pluginScore.Score)
```

File: pkg/scheduler/schedule_one.go [hunk 14]
```diff
@@ -692,16 +699,16 @@ func prioritizeNodes(
...
 				}
 				mu.Lock()
 				defer mu.Unlock()
 				for i := range *prioritizedList {
 					nodename := (*prioritizedList)[i].Host
 					score := (*prioritizedList)[i].Score
-					if klogV.Enabled() {
-						klogV.InfoS("Extender scored node for pod", "pod", klog.KObj(pod), "extender", extenders[extIndex].Name(), "node", nodename, "score", score)
+					if loggerVTen.Enabled() {
+						logger.V(10).Info("Extender scored node for pod", "pod", klog.KObj(pod), "extender", extenders[extIndex].Name(), "node", nodename, "score", score)
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:711]:
```suggestion
						loggerVTen.Info("Extender scored node for pod", "pod", klog.KObj(pod), "extender", extenders[extIndex].Name(), "node", nodename, "score", score)
```

File: pkg/scheduler/schedule_one.go [hunk 15]
```diff
@@ -732,9 +739,9 @@ func prioritizeNodes(
 		}
 	}
 
-	if klogV.Enabled() {
+	if loggerVTen.Enabled() {
 		for i := range nodesScores {
-			klogV.InfoS("Calculated node's final score for pod", "pod", klog.KObj(pod), "node", nodesScores[i].Name, "score", nodesScores[i].TotalScore)
+			logger.V(10).Info("Calculated node's final score for pod", "pod", klog.KObj(pod), "node", nodesScores[i].Name, "score", nodesScores[i].TotalScore)
```
Comment (alculquicondor) [inline pkg/scheduler/schedule_one.go:744]:
```suggestion
			loggerVTen.Info("Calculated node's final score for pod", "pod", klog.KObj(pod), "node", nodesScores[i].Name, "score", nodesScores[i].TotalScore)
```

File: pkg/scheduler/internal/cache/cache_test.go (hunk 7 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/cache_test.go:331]: For the original PR, this test has been changed in PR https://github.com/kubernetes/kubernetes/pull/110954 and I didn't notice the change. 
Update to the latest master branch. Thanks for the catch!

File: pkg/scheduler/internal/cache/cache_test.go (hunk 8 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/cache_test.go:395]: Yes. I will open a new PR to follow up.

File: pkg/scheduler/schedule_one.go (hunk 12 above)
Comment (mengjiao-liu) [reply pkg/scheduler/schedule_one.go:159]: Oh... sorry. 
The previous PR #111155  error changed this line, and I didn't notice this.
Changed this to `loggerV(5).InfoS`

File: pkg/scheduler/internal/cache/node_tree.go (hunk 9 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/node_tree.go]: Done.

File: pkg/scheduler/schedule_one.go (hunk 14 above)
Comment (mengjiao-liu) [reply pkg/scheduler/schedule_one.go]: Done.

File: pkg/scheduler/schedule_one.go (hunk 15 above)
Comment (mengjiao-liu) [reply pkg/scheduler/schedule_one.go]: Done.

File: pkg/scheduler/internal/cache/node_tree.go (hunk 10 above)
Comment (mengjiao-liu) [reply pkg/scheduler/internal/cache/node_tree.go]: Done.

File: pkg/scheduler/schedule_one.go (hunk 11 above)
Comment (mengjiao-liu) [reply pkg/scheduler/schedule_one.go]: > Is there an issue for this? If so, please reference TODO(#1234)

Yes. I have added the issue number.

> Also, does it need to be a context or can it be a logger?

As far as I know, the function `utilruntime.HandleError` is used in many components. 
Do we have plans to decide whether to pass context or logger in the future? @pohly

Comment (alculquicondor) [reply pkg/scheduler/schedule_one.go]: Can you open an issue to add a logger of Ctx to `HandleError`?

**Review by hashim21223445 — APPROVED** [review]
